### PR TITLE
feat(opencode): trigger compaction earlier and add multi-file read

### LINKED ISSUE #9045: Codex model response latency increases significantly as conversation grows
## Problem

When using Codex models (via OAuth authentication), response latency increases significantly as the conversation grows longer. This happens because the entire conversation history is sent with every API request.

## Root Cause

After investigating the codebase:

1. The OpenAI Responses API returns a `responseId` in `providerMetadata.openai.responseId`
2. The SDK supports `previousResponseId` option (`openai-responses-language-model.ts:285`)
3. However, **this feature is never actually used** - the `responseId` from responses is not stored, and `previousResponseId` is never set

### Current Flow
```
Request 1 → Send full history → Get response with responseId (discarded)
Request 2 → Send full history again → Get response with responseId (discarded)
Request N → Send increasingly large history → Slow response
```

### Expected Flow with previous_response_id
```
Request 1 → Send full history → Get response with responseId (saved)
Request 2 → Send previousResponseId + new message only → Fast response
Request N → Reference previous response → Consistent speed
```

## Impact

- Codex API calls become progressively slower as conversations grow
- Poor user experience with long coding sessions
- Unnecessary bandwidth and compute usage

## Proposed Solution

1. Add `responseId` field to `AssistantMessage` schema
2. Store `responseId` from `providerMetadata` when receiving responses
3. Pass `previousResponseId` to subsequent requests via `providerOptions`

### PR DESCRIPTION
**Pull request by Skyline-23**
## Summary
- trigger session compaction earlier to avoid overflow at high context usage
- add multi-file read validation and update read tool guidance
- expand read and compaction tests to cover filePaths edge cases and early threshold

## Testing
- `bun test test/tool/read.test.ts`
- `bun test test/session/compaction.test.ts`

## References
- Fixes #9045

### CODE REVIEW

**Comment by Skyline-23** [comment]
Updated the PR title to conventional format and added `Fixes #9045` in the description. Addressed review feedback with filePaths validation, docs updates, added edge-case tests, Promise.all multi-read, and early compaction threshold coverage.

File: packages/opencode/test/tool/read.test.ts [hunk 1]
```diff
@@ -51,6 +51,29 @@ describe("tool.read external_directory permission", () => {
...
+      fn: async () => {
+        const read = await ReadTool.init()
+        const result = await read.execute(
+          {
+            filePaths: [path.join(tmp.path, "first.txt"), path.join(tmp.path, "second.txt")],
+          },
+          ctx,
+        )
+        expect(result.output).toContain("first content")
+        expect(result.output).toContain("second content")
```
[the remark this answers is not in the document]
Comment (Skyline-23) [reply packages/opencode/test/tool/read.test.ts]: Added filePaths tests to assert `<file>` blocks, output order, and multi-file title formatting.

File: packages/opencode/test/tool/read.test.ts [hunk 2]
```diff
@@ -51,6 +51,29 @@ describe("tool.read external_directory permission", () => {
...
+      },
+    })
+  })
```
[the remark this answers is not in the document]
Comment (Skyline-23) [reply packages/opencode/test/tool/read.test.ts]: Added edge-case tests for missing file, offset/limit per file, image/PDF attachments, and empty filePaths array.

File: packages/opencode/src/tool/read.ts [hunk 3]
```diff
@@ -15,123 +15,166 @@ const MAX_BYTES = 50 * 1024
...
+  parameters: z
+    .object({
+      filePath: z.string().describe("The path to the file to read").optional(),
+      filePaths: z.array(z.string()).describe("Paths to files to read").optional(),
+      offset: z.coerce.number().describe("The line number to start reading from (0-based)").optional(),
+      limit: z.coerce.number().describe("The number of lines to read (defaults to 2000)").optional(),
+    })
+    .refine((value) => value.filePath || (value.filePaths && value.filePaths.length > 0), {
+      message: "filePath or filePaths is required",
+    }),
```
[the remark this answers is not in the document]
Comment (Skyline-23) [reply packages/opencode/src/tool/read.ts:30]: Added schema refinement plus a runtime guard to enforce `filePath` XOR `filePaths`, with a clear error message.

File: packages/opencode/src/tool/read.ts [hunk 4]
```diff
@@ -15,123 +15,166 @@ const MAX_BYTES = 50 * 1024
...
-    filePath: z.string().describe("The path to the file to read"),
-    offset: z.coerce.number().describe("The line number to start reading from (0-based)").optional(),
-    limit: z.coerce.number().describe("The number of lines to read (defaults to 2000)").optional(),
-  }),
+  parameters: z
+    .object({
+      filePath: z.string().describe("The path to the file to read").optional(),
+      filePaths: z.array(z.string()).describe("Paths to files to read").optional(),
+      offset: z.coerce.number().describe("The line number to start reading from (0-based)").optional(),
+      limit: z.coerce.number().describe("The number of lines to read (defaults to 2000)").optional(),
```
[the remark this answers is not in the document]
Comment (Skyline-23) [reply packages/opencode/src/tool/read.ts:23]: Documented that `offset`/`limit` apply to each file when using `filePaths` in the read tool docs.

File: packages/opencode/src/tool/read.ts [hunk 5]
```diff
@@ -15,123 +15,166 @@ const MAX_BYTES = 50 * 1024
...
-    if (truncatedByBytes) {
-      output += `\n\n(Output truncated at ${MAX_BYTES} bytes. Use 'offset' parameter to read beyond line ${lastReadLine})`
-    } else if (hasMoreLines) {
-      output += `\n\n(File has more lines. Use 'offset' parameter to read beyond line ${lastReadLine})`
-    } else {
-      output += `\n\n(End of file - total ${totalLines} lines)`
+    const results = []
+    for (const value of paths) {
+      results.push(await readOne(value))
     }
```
[the remark this answers is not in the document]
Comment (Skyline-23) [reply packages/opencode/src/tool/read.ts]: Switched multi-file reads to `Promise.all` to run reads in parallel.

File: packages/opencode/src/session/compaction.ts [hunk 6]
```diff
@@ -35,11 +35,13 @@ export namespace SessionCompaction {
...
     const output = Math.min(input.model.limit.output, SessionPrompt.OUTPUT_TOKEN_MAX) || SessionPrompt.OUTPUT_TOKEN_MAX
     const usable = input.model.limit.input || context - output
-    return count > usable
+    const threshold = Math.floor(usable * EARLY_COMPACT_RATIO)
+    return count > threshold
   }
 
   export const PRUNE_MINIMUM = 20_000
   export const PRUNE_PROTECT = 40_000
+  export const EARLY_COMPACT_RATIO = 0.9
```
[the remark this answers is not in the document]
Comment (Skyline-23) [reply packages/opencode/src/session/compaction.ts:44]: Added an explicit test that verifies the early compaction threshold (90% of usable context) triggers overflow.
